sbi/inference/npse/npse.py

```diff
@@ -0,0 +1,612 @@
+
+        Returns:
+            Posterior $p(\theta|x)$  with `.sample()` and `.log_prob()` methods
+            (the returned log-probability is unnormalized).
+        """
+        if prior is None:
+            assert self._prior is not None, (
+                "You did not pass a prior. You have to pass the prior either at "
+                "initialization `inference = NPSE(prior)` or to "
+                "`.build_posterior(prior=prior)`."
+            )
+            prior = self._prior
+        else:
+            utils.check_prior(prior)
+
+        if score_estimator is None:
+            score_estimator = self._neural_net
+            # If internal net is used device is defined.
+            device = self._device
+        else:
+            # TODO: Add protocol for checking if the score estimator has forward and
+            # loss methods with the correct signature.
+            score_estimator = score_estimator
+            # Otherwise, infer it from the device of the net parameters.
+            device = str(next(score_estimator.parameters()).device)
+
+        posterior = ScorePosterior(
+            score_estimator,  # type: ignore
+            prior,
+            device=device,
+            sample_with=sample_with,
+        )
+
+        self._posterior = posterior
+        # Store models at end of each round.
+        self._model_bank.append(deepcopy(self._posterior))
+
+        return deepcopy(self._posterior)
+
+    def _loss_proposal_posterior(
+        self,
+        theta: Tensor,
+        x: Tensor,
+        masks: Tensor,
+        proposal: Optional[Any],
+    ) -> Tensor:
+        raise NotImplementedError("Multi-round NPSE is not yet implemented.")
+
+    def _loss(
+        self,
+        theta: Tensor,
+        x: Tensor,
+        masks: Tensor,
+        proposal: Optional[Any],
+        calibration_kernel: Callable,
+        force_first_round_loss: bool = False,
+    ) -> Tensor:
+        """Return loss from score estimator. Currently only single-round NPSE
+         is implemented, i.e., no proposal correction is applied for later rounds.
+
+        The loss is the negative log prob. Irrespective of the round or SNPE method
+        (A, B, or C), it can be weighted with a calibration kernel.
+
+        Returns:
+            Calibration kernel-weighted negative log prob.
+            force_first_round_loss: If `True`, train with maximum likelihood,
+                i.e., potentially ignoring the correction for using a proposal
+                distribution different from the prior.
+        """
+        if self._round == 0 or force_first_round_loss:
+            # First round loss.
+            loss = self._neural_net.loss(theta, x)
+        else:
+            # TODO: Implement proposal correction for multi-round SNPSE.
+            loss = self._loss_proposal_posterior(theta, x, masks, proposal)
+
+        return calibration_kernel(x) * loss
+
+    def _check_proposal(self, proposal):
```

Comment:
I would remove this, and then also remove the option to pass a proposal to `append_simulations` and remove the `check_proposal` there.